File: vol2/ch21/LeetCode10.cpp
class Solution {
public:
	bool isMatch(string &s,string &p) {
     	int m=s.size(),n=p.size();
        	auto dp=vector<vector<bool>>(m+1,vector<bool>(n+1,false));
        	dp[0][0]=true;
        	for(int j=1;j<=n;j++) { 
          	if(p[j-1]=='*' && j-2>=0 && dp[0][j-2]) 
              	dp[0][j]=true;
        	}
        	for (int i=1;i<=m;i++) {
            	for (int j=1;j<=n;j++) {
                	if(p[j-1]!='*') {
                    	if (p[j-1]=='.'  || s[i-1]==p[j-1])
                        	dp[i][j]=dp[i-1][j-1];
                	}
				else {										// p[j-1]=='*'
                   	dp[i][j]=dp[i][j-2];
                    	if (p[j-2]=='.'  || s[i-1]==p[j-2])
                        	dp[i][j]=dp[i][j] || dp[i-1][j];
                	}
            	}
        	}
        	return dp[m][n];
    	}
};

File: vol2/ch21/LeetCode10.py
class Solution:
  def isMatch(self, s: str, p: str) -> bool:
    m,n=len(s),len(p)
    dp=[[False for _ in range(n+1)] for _ in range(m+1)]
    dp[0][0]=True
    for j in range(1,n+1): 
      if p[j-1]=='*' and j-2>=0 and dp[0][j-2]: 
        dp[0][j]=True
    for i in range(1,m+1):
      for j in range(1,n+1):
        if p[j-1]!='*':
          if p[j-1]=='.' or s[i-1]==p[j-1]:
            dp[i][j]=dp[i-1][j-1]
        else:     #p[j-1]=='*'
          dp[i][j]=dp[i][j-2]
          if p[j-2]=='.'  or s[i-1]==p[j-2]:
            dp[i][j]=dp[i][j] or dp[i-1][j]
    return dp[m][n]
